Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,188 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "config.h"
+#import "WebExtensionContext.h"
+
+#import "APIContentWorld.h"
+#import "CocoaHelpers.h"
+#import "WKContentWorld.h"
+#import "WKFrameInfoPrivate.h"
+#import "WKWebViewPrivate.h"
+#import "WebExtensionContextProxy.h"
+#import "WebExtensionContextProxyMessages.h"
+#import "WebExtensionScriptInjectionParameters.h"
+#import "WebExtensionScriptInjectionResultParameters.h"
+#import "WebExtensionDynamicScripts.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionTabIdentifier.h"
+#import "WebExtensionUtilities.h"
+#import "WebPageProxy.h"
+#import "WebProcessProxy.h"
+#import "WebUserContentControllerProxy.h"
+#import "_WKFrameHandle.h"
+#import "_WKFrameTreeNode.h"
+#import <WebCore/UserStyleSheet.h>
+#import <wtf/BlockPtr.h>
+#import <wtf/cocoa/VectorCocoa.h>
+
+namespace WebKit {
+using namespace DynamicScripts;
+
+using SourcePair = DynamicScripts::SourcePair;
+using ScriptVector = DynamicScripts::StyleSheetPairs;
+
+void WebExtensionContext::scriptingExecuteScript(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
+{
+    WKWebView *webView = m_tabMap.get(parameters.identifier.value())->mainWebView();
+    if (!hasPermission(webView.URL)) {
+        callback(std::nullopt);
+        return;
+    }
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, parameters = WTFMove(parameters), callback = WTFMove(callback)](_WKFrameTreeNode *mainFrame) mutable {
+        API::ContentWorld& executionWorld = *m_contentScriptWorld;
+        if (parameters.world == WebExtensionContentWorldType::Main)
+            executionWorld = API::ContentWorld::pageContentWorld();
+
+        Vector<RetainPtr<_WKFrameTreeNode>> frames;
+        if (parameters.allFrames || !parameters.frameIDs)
+            populateFrames(mainFrame, mainFrameID, parameters.allFrames, false, WTFMove(frames));
+        else
+            populateFrames(mainFrame, parameters.frameIDs.value(), WTFMove(frames));
+
+        size_t numberOfExecutions = frames.size();
+        Vector<WebExtensionScriptInjectionResultParameters> results;
+
+        for (auto& frame : frames) {
```

Comment:
This seems wrong. Didn't you `WTFMove` the frames variable into `populateFrames`? That should mean that it is now empty. Whatever happens to `frames` inside of `populateFrames` will be invisible up at this layer of the stack. I think `frames` should just be passed as a reference so that we can observe its state after the call to `populateFrames`.

It actually might be a bit clearer if `populateFrames` just returned the populated Vector, since you don't seem to ever want to pass frames from elsewhere into that method.